File: src/main/java/skywolf46/extrautility/util/TimeParser.java
package skywolf46.extrautility.util;

import java.util.HashMap;
import java.util.function.Function;

public class TimeParser {
    private static ParsingType parser;

    static {
        parser = new ParsingType();
        parser.append("ms", Integer::parseInt);
        parser.append("s", x -> Integer.parseInt(x) * 1000);
        parser.append("m", x -> Integer.parseInt(x) * 60 * 1000);
        parser.append("h", x -> Integer.parseInt(x) * 60 * 60 * 1000);
        parser.append("d", x -> Integer.parseInt(x) * 24 * 60 * 60 * 1000);
    }

    public static long parseToMillisecond(String x, ParsingType parser) {
        long total = 0;
        int toParse = 0;
        int lastParsed = 0;
        for (int i = 0; i < x.length(); i++) {
            char cx = x.charAt(i);
            if (CharacterUtil.isDigit(cx)) {
                if (toParse != lastParsed) {
                    long am = parse(x, x.substring(lastParsed, toParse + 1), parser);
                    if (am != Integer.MIN_VALUE) {
                        total += am;
                        toParse = i;
                        lastParsed = i;
                    }
                }
            } else {
                toParse = i;
            }
        }
        if (toParse != lastParsed) {
            long am = parse(x, x.substring(lastParsed), parser);
            if (am != Integer.MIN_VALUE)
                total += am;
        }
        return total;
    }

    public static long parseToMillisecond(String x) {
        return parseToMillisecond(x, parser);
    }

    private static long parse(String original, String toParse, ParsingType parser) {
        String target = null;
        for (int i = 0; i < toParse.length(); i++) {
            if (!CharacterUtil.isDigit(toParse.charAt(i))) {
                target = toParse.substring(0, i);
                toParse = toParse.substring(i);
                break;
            }
        }
        if (target == null)
            throw new IllegalStateException("Time is not number");
        if (!parser.canAccept(toParse))
            throw new IllegalStateException("Cannot parse \"" + toParse + "\" at \"" + original + "\" : Not a parsing target");
        Function<String, Integer> fi = parser.parser(toParse);
        if (fi == null)
            return Integer.MIN_VALUE;
        return fi.apply(target);
    }


    public static class ParsingType {
        private HashMap<String, Function<String, Integer>> parsers = new HashMap<>();

        private int maxLength = 0;

        public ParsingType append(String type, Function<String, Integer> funct) {
            parsers.put(type, funct);
            maxLength = Math.max(type.length(), maxLength);
            return this;
        }

        public int getMaxLength() {
            return maxLength;
        }

        public Function<String, Integer> parser(String x) {
            return parsers.get(x);
        }

        public boolean canAccept(String toParse) {
            return toParse.length() <= maxLength;
        }
    }

}

File: common/src/main/kotlin/skywolf46/extrautility/util/TimeParser.kt
package skywolf46.extrautility.util

import java.util.function.Function
import kotlin.math.max


object TimeParser {
    private var parser: ParsingType? = null

    @JvmOverloads
    fun parseToMillisecond(x: String, parser: ParsingType? = TimeParser.parser): Long {
        var total: Long = 0
        var toParse = 0
        var lastParsed = 0
        for (i in x.indices) {
            val cx = x[i]
            if (cx.isDigit()) {
                if (toParse != lastParsed) {
                    val am = parse(x, x.substring(lastParsed, toParse + 1), parser)
                    if (am != Int.MIN_VALUE.toLong()) {
                        total += am
                        toParse = i
                        lastParsed = i
                    }
                }
            } else {
                toParse = i
            }
        }
        if (toParse != lastParsed) {
            val am = parse(x, x.substring(lastParsed), parser)
            if (am != Int.MIN_VALUE.toLong()) total += am
        }
        return total
    }

    private fun parse(original: String, toParse: String, parser: ParsingType?): Long {
        var toParse = toParse
        var target: String? = null
        for (i in toParse.indices) {
            if (!toParse[i].isDigit()) {
                target = toParse.substring(0, i)
                toParse = toParse.substring(i)
                break
            }
        }
        checkNotNull(target) { "Time is not number" }
        check(parser!!.canAccept(toParse)) { "Cannot parse \"$toParse\" at \"$original\" : Not a parsing target" }
        val fi = parser.parser(toParse)
            ?: return Int.MIN_VALUE.toLong()
        return fi.apply(target).toLong()
    }

    class ParsingType {
        private val parsers = HashMap<String, Function<String, Int>>()
        private var maxLength = 0

        fun append(type: String, funct: Function<String, Int>): ParsingType {
            parsers[type] = funct
            maxLength = max(type.length, maxLength)
            return this
        }

        fun parser(x: String): Function<String, Int>? {
            return parsers[x]
        }

        fun canAccept(toParse: String): Boolean {
            return toParse.length <= maxLength
        }
    }

    init {
        parser = ParsingType()
        parser!!.append("ms") { s: String -> s.toInt() }
        parser!!.append("s") { x: String -> x.toInt() * 1000 }
        parser!!.append("m") { x: String -> x.toInt() * 60 * 1000 }
        parser!!.append("h") { x: String -> x.toInt() * 60 * 60 * 1000 }
        parser!!.append("d") { x: String -> x.toInt() * 24 * 60 * 60 * 1000 }
    }
}